File: 23.tell-dont-ask/src/java/src/main/java/com/telldontask/usecase/OrderApprovalUseCase.java
package com.telldontask.usecase;

import com.telldontask.domain.Order;
import com.telldontask.domain.OrderStatus;
import com.telldontask.repository.OrderRepository;

public class OrderApprovalUseCase {
    private final OrderRepository orderRepository;

    public OrderApprovalUseCase(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public void run(OrderApprovalRequest request) {
        Order order = orderRepository.getById(request.getOrderId());

        if (order.getStatus() == OrderStatus.SHIPPED) {
            throw new ShippedOrdersCannotBeChangedException();
        }

        if (request.isApproved() && order.getStatus() == OrderStatus.REJECTED) {
            throw new RejectedOrderCannotBeApprovedException();
        }

        if (!request.isApproved() && order.getStatus() == OrderStatus.APPROVED) {
            throw new ApprovedOrderCannotBeRejectedException();
        }

        order.setStatus(request.isApproved() ? OrderStatus.APPROVED : OrderStatus.REJECTED);
        orderRepository.save(order);
    }
}

File: 23.tell-dont-ask/src/typescript/useCase/orderApprovalUseCase.ts
import { OrderRepository } from '../repository/orderRepository';
import { OrderStatus } from '../domain/orderStatus';
import { OrderApprovalRequest } from './orderApprovalRequest';
import { ShippedOrdersCannotBeChangedException } from './shippedOrdersCannotBeChangedException';
import { RejectedOrderCannotBeApprovedException } from './rejectedOrderCannotBeApprovedException';
import { ApprovedOrderCannotBeRejectedException } from './approvedOrderCannotBeRejectedException';

export class OrderApprovalUseCase {
  constructor(private readonly orderRepository: OrderRepository) {}

  run(request: OrderApprovalRequest): void {
    const order = this.orderRepository.getById(request.orderId)!;

    if (order.status === OrderStatus.Shipped) {
      throw new ShippedOrdersCannotBeChangedException();
    }

    if (request.approved && order.status === OrderStatus.Rejected) {
      throw new RejectedOrderCannotBeApprovedException();
    }

    if (!request.approved && order.status === OrderStatus.Approved) {
      throw new ApprovedOrderCannotBeRejectedException();
    }

    order.status = request.approved ? OrderStatus.Approved : OrderStatus.Rejected;
    this.orderRepository.save(order);
  }
}
